### PR TITLE
[SPARK-10031][SQL] Join two UnsafeRows in SortMergeJoin if possible

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-10031

Currently in `SortMergeJoin`, when two rows from left and right plans are both `UnsafeRow`, we still use `JoinedRow` to join them and do an extra `UnsafeProjection` later.

We can just use `GenerateUnsafeRowJoiner` to join two `UnsafeRows` in `SortMergeJoin` if possible. Besides, `GenerateUnsafeRowJoiner` can have a `withRight` function to only update row2 with a same row1.

Add `withRight` function to `GenerateUnsafeRowJoiner`. We can avoid another unnecessary data copying for row1.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala [hunk 1]
```diff
@@ -96,9 +103,37 @@ case class SortMergeJoin(
...
+            }
+          } else {
+            rowJoiner.withRight
+          }
+        }
+
         private[this] val resultProjection: (InternalRow) => InternalRow = {
-          if (isUnsafeMode) {
+          if (canUnsafeJoin) {
+            identity[InternalRow]
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala]: Two `UnsafeRows` are joined together as another `UnsafeRow`. So we don't need to do `UnsafeProjection` here.

**Comment by viirya** [comment]
retest this please.

**Comment by chenghao-intel** [comment]
As we cannot tell what downstream operator is inside of the spark plan operator, it's hard to tell if it's really an optimization. JoinedRow is used here for reducing the memory copying or object creation , but this PR does creates small objects and copying memory when constructing new UnsafeRow, right?

I'll agree if we can move the logic into `SparkStrategy`, what do you think?

**Comment by viirya** [comment]
@chenghao-intel Currently we will do an `UnsafeProjection` for two `UnsafeRow`s. I think it will also create a new `UnsafeRow`?

This patch replaces it with a `GenerateUnsafeRowJoiner` to join two `UnsafeRow`s. For `InternalRow`s, this patch does as same as the current way (`JoinedRow`).

This patch also optimises `GenerateUnsafeRowJoiner` with a new `withRight` method. So we can skip some unnecessary memory copying.

**Comment by chenghao-intel** [comment]
I agree this is an improvement with the original code. However, I think the original code is probably a little over in optimization.

Take the simple query for example:
```explain extended select * from src1 a join src2 b on a.key=b.key and a.value > b.value;```
And the physical plan like:
···
TungstenProject [key#11,value#12,key#13,value#14]
 Filter (value#12 > value#14)
  SortMergeJoin [key#11], [key#13]
   TungstenSort [key#11 ASC], false, 0
    TungstenExchange hashpartitioning(key#11)
     ConvertToUnsafe
      HiveTableScan [key#11,value#12], (MetastoreRelation default, src1, Some(a))
   TungstenSort [key#13 ASC], false, 0
    TungstenExchange hashpartitioning(key#13)
     ConvertToUnsafe
      HiveTableScan [key#13,value#14], (MetastoreRelation default, src2, Some(b))
···

We will create many small objects (and memory copying) in `SortMergeJoin`, but probably only few records will be filtered in the `Filter`, a normal `JoinedRow` probably more efficient in this case.

cc @JoshRosen

**Comment by andrewor14** [comment]
@JoshRosen any thoughts?

**Comment by davies** [comment]
Currently, We can't return a JoinedRow with two UnsafeRow in it (because JoinedRow.getUTFString() does not copy the underlying bytes, also for complex types). 

As @chenghao-intel said, switching to GenerateUnsafeRowJoiner could improve the copy, but actually the copy could be avoided at all. For example, we could have a UnsafeJoinedRow, which is similar to JoinedRow, but also have some features as UnsafeRow (fast serialization).

Another direction we could explore is that we could compile the whole query plan into one function to avoid the materialization of rows, that will be different story.

**Comment by andrewor14** [comment]
@viirya any updates on this? Is this still actively worked on?

**Comment by viirya** [comment]
@andrewor14 I will try to update this soon. By considering @davies's suggestion, I think if we can add an `UnsafeJoinedRow` for this, instead of using `GenerateUnsafeRowJoiner` now.

**Comment by andrewor14** [comment]
Let's close this PR for now because it's mostly gone stale and there hasn't been any activity for 2 months.

**Comment by viirya** [comment]
Ok. Close it now.

**Pull request closed without merging**